ads/adocean.js

```diff
@@ -178,15 +317,35 @@ export function adocean(global, data) {
     'aoKeys',
     'aoVars',
     'aoClusters',
+    'aoMaster',
   ]);
 
-  const mode = (data['aoMode'] != 'sync') ? 'buffered' : 'sync';
+  const masterId = data['aoMaster'];
+  const mode = (data['aoMode'] !== 'sync' || masterId ? 'buffered' : 'sync');
   const adoUrl = 'https://' + data['aoEmitter'] + ADO_JS_PATHS[mode];
 
+  /*
+   * INSUFFICIENT and UNKOWN should be threated as INSUFFICIENT
+   * not defined states should be threated as INSUFFICIENT
+   */
+  const consent = (global.context.initialConsentState === null /* tags without data-block-on-consent */
+                   || global.context.initialConsentState === CONSENT_POLICY_STATE.SUFFICIENT
+                   || global.context.initialConsentState === CONSENT_POLICY_STATE.UNKNOWN_NOT_REQUIRED);
+
   writeScript(global, adoUrl, () => {
-    setupAdoConfig(mode, global);
+    setupAdoConfig(mode, global, consent);
     setupPreview(global, data);
 
-    appendPlacement(mode, global, data);
+    if (masterId) {
+      if (global['ado']['features'] && global['ado']['features']['passback']) {
```

Comment:
is `global['ado']` always guaranteed to be defined?

Reply:
It's defined inside external script, which can failed to load. I'll add additional check for that.

Reply:
It's fixed in https://github.com/ampproject/amphtml/pull/21328